files/migrations.py

```diff
@@ -0,0 +1,188 @@
+        """
+        chunks = []
+        # Opting for minimal code over execution speed
+        for part in version.split("."):
+            digits = ""
+            for char in part:
+                if char.isdigit():
+                    digits += char
+                else:
+                    # Treat RCs, betas etc. as if it were the full release
+                    break
+            chunks.append(int(digits))
+        self.version = tuple(chunks)
+
+    def __str__(self):
+        return ".".join(str(v) for v in self.version)
+
+
+class Migration(Version):
+    """
+    A sortable runnable migration
+
+    Subclass of Version so that we can sort migrations just as easily as version strings
+    """
+
+    def __init__(self, path):
+        super().__init__(path.name.rsplit(".", 1)[0])
+        self.path = path
+
+    def run_and_update_version_file(self, version_file):
+        """
+        Run this migration
+        """
+
+        try:
+            loader = SourceFileLoader("migration", str(self.path))
+            spec = spec_from_loader("migration", loader)
+            migration = module_from_spec(spec)
+            loader.exec_module(migration)
+            log.info(f"Running migration for {self}")
+            migrate(migration.steps)
+        except Exception as error:
+            log.error(f"{error}")
+            log.error(f"Failed to load migration: {self.path}")
+            exit(2)
+
+        # If we successfully ran the migration, we are now in a new state that we want to see
+        # reflected in case a subsequent migration fails
+        update_version(self, version_file)
+
+
+def update_version(target, version_file):
+    version_file.write_text(f"{target}")
+
+
+def _rollback_steps(steps, failed_index, tmpdir):
+    for i, step in reversed(list(enumerate(steps[:failed_index]))):
+        try:
+            step.rollback(tmpdir / f"{step.__class__.__name__}{i}")
+        except Exception as error:
+            log.error(f"{error}")
+            log.error(f"Failed to roll back step {i}: {step}")
+            pass
+
+
+def _validate(steps, tmpdir):
+    # Before attempting to run anything, ensure that the required state is met
+    for i, step in enumerate(steps):
+        try:
+            if not step.validate():
+                log.error(f"Failed to validate step {i}: {step}")
+                exit(2)
+        except Exception as error:
+            log.error(f"{error}")
+            log.error(f"Encountered error during validation step {i}: {step}")
+            exit(2)
+
+
+def _snapshot(steps, tmpdir):
+    for i, step in enumerate(steps):
+        tmpdir_step = tmpdir / f"{step.__class__.__name__}{i}"
+        tmpdir_step.mkdir()
+        try:
+            step.snapshot(tmpdir_step)
+        except Exception as error:
+            log.error(f"{error}")
+            _rollback_steps(steps, i, tmpdir)
+            log.error(f"Failed to snapshot step {i}: {step}")
```

Comment:
This seems like it should go before the rollback attempt?

Reply:
Upon further inspection, I think a rollback attempt here actually doesn't make sense, so I removed it. Nothing has been run yet, the snapshot stage is separate, and rollback is meant to undo what run tried to do. But run hasn't done anything yet, so rollback shouldn't even happen. Let me know if you think I got that wrong though!

Reply:
Yep, that makes sense.